### PR TITLE
DAOS-10365 dfuse: Allow dfuse caching of attributes.

### PR DESCRIPTION
**Pull request by ashleypittman**
Use cached values for inodes when required.
When the kernel is using writeback cache then it can query inode
metadata before the timeout has expired and this costs every time
a file is opened.  To handle this keep a record of when the inode
was last reported to the kernel and for requests before the
timeout then reply with the cached data, but use a shorter timeout.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>
# ------------------------ >8 ------------------------
Skip-func-hw-test: true
Skip-func-test: true
Quick-Functional: true
Test-tag: dfuse

### CODE REVIEW

**Comment by mchaarawi** [comment]
'''src/client/dfuse/ops/setattr.c: In function ‘dfuse_cb_setattr’:
src/client/dfuse/ops/setattr.c:115:14: error: incompatible types when assigning to type ‘struct stat’ from type ‘struct stat *’
  ie->ie_stat = attr;
              ^
'''

File: src/client/dfuse/ops/setattr.c [hunk 1]
```diff
@@ -112,11 +112,7 @@ dfuse_cb_setattr(fuse_req_t req, struct dfuse_inode_entry *ie,
 
 	attr->st_ino = ie->ie_stat.st_ino;
 
-	/* Update the size as dfuse knows about it for future use, but only if it was set as part
-	 * of this call.  See DAOS-8333
-	 */
-	if (dfs_flags & DFS_SET_ATTR_SIZE)
-		ie->ie_stat.st_size = attr->st_size;
+	ie->ie_stat = attr;
```
Comment (mchaarawi) [inline src/client/dfuse/ops/setattr.c:115]: you mean
memcpy(&ie->ie_stat, attr, sizeof(struct stat));
?

Comment (ashleypittman) [reply src/client/dfuse/ops/setattr.c]: I was trying to be lazy.

**Comment by ashleypittman** [comment]
@mchaarawi what do you think about this one?  It will need updating to drop the cache on writes as the size may change but it should be safe against setattr calls as they reset the cache anyway.  I need to verify the locking if we're going to move forward with it.

The same cache-timeout logic is also needed for the readdir cache where the kernel does not keep track of age.

**Comment by mchaarawi** [comment]
> @mchaarawi what do you think about this one? It will need updating to drop the cache on writes as the size may change but it should be safe against setattr calls as they reset the cache anyway. I need to verify the locking if we're going to move forward with it.
> 
> The same cache-timeout logic is also needed for the readdir cache where the kernel does not keep track of age.

yea i think this was pretty important as the getattr was pretty expensive on the ai benchmark. the patch looks good to me. although i would like to see something similar like the ai benchmark was doing in a unit test at smaller scale and to verify the getattr gets called once.

**Review by mchaarawi — APPROVED** [review]

**Comment by ashleypittman** [comment]
> > @mchaarawi what do you think about this one? It will need updating to drop the cache on writes as the size may change but it should be safe against setattr calls as they reset the cache anyway. I need to verify the locking if we're going to move forward with it.
> > The same cache-timeout logic is also needed for the readdir cache where the kernel does not keep track of age.
> 
> yea i think this was pretty important as the getattr was pretty expensive on the ai benchmark. the patch looks good to me. although i would like to see something similar like the ai benchmark was doing in a unit test at smaller scale and to verify the getattr gets called once with the timeout set to something large.

As I said it needs extra code to evict the cache on writes so I don't think it's ready to land yet.  Perhaps @phender or @jamesanunez have ideas for how we could test if this is working.  I can fairly easily write a unit-test to invoke it but we then need to look at the log/statistics to see that it's working.

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**